File: algorithms/Heap Sort/C/code.c
#include <stdio.h>

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

void heapify(int arr[], int size, int root)
{
    int largest = root;
    int left = 2 * root + 1;
    int right = 2 * root + 2;

    if (left < size && arr[left] > arr[largest])
    {
        largest = left;
    }

    if (right < size && arr[right] > arr[largest])
    {
        largest = right;
    }

    if (largest != root)
    {
        swap(&arr[root], &arr[largest]);

        heapify(arr, size, largest);
    }
}

void heap_sort(int arr[], int size)
{
    /*
     * Build the max heap.
     */
    for (int i = size / 2 - 1; i >= 0; i--)
    {
        heapify(arr, size, i);
    }

    /*
     * Move the largest element to the end
     * and restore the heap.
     */
    for (int i = size - 1; i > 0; i--)
    {
        swap(&arr[0], &arr[i]);

        heapify(arr, i, 0);
    }
}

int main(void)
{
    int arr[] = {12, 11, 13, 5, 6, 7};
    int size = sizeof(arr) / sizeof(arr[0]);

    heap_sort(arr, size);

    for (int i = 0; i < size; i++)
    {
        printf("%d ", arr[i]);
    }

    printf("\n");

    return 0;
}

File: algorithms/Heap Sort/Python/code.py
def heap_sort(arr):
    def heapify(size, root):
        largest = root
        left = 2 * root + 1
        right = 2 * root + 2

        if left < size and arr[left] > arr[largest]:
            largest = left

        if right < size and arr[right] > arr[largest]:
            largest = right

        if largest != root:
            arr[root], arr[largest] = arr[largest], arr[root]

            heapify(size, largest)

    size = len(arr)

    # Build the max heap.
    for i in range(size // 2 - 1, -1, -1):
        heapify(size, i)

    # Move the largest element to the end.
    for i in range(size - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]

        # Restore the max-heap property.
        heapify(i, 0)


arr = [12, 11, 13, 5, 6, 7]

heap_sort(arr)

print(arr)